README.md
plugin-digicode


fr_FR/changelog.md
### Version 1.0 (version market 2018-08-19 22:01:25)
- Verion initiale


plugin_info/install.php
<?php

/* This file is part of Jeedom.
 *
 * Jeedom is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jeedom is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jeedom. If not, see <http://www.gnu.org/licenses/>.
 */

require_once dirname(__FILE__) . '/../../../core/php/core.inc.php';

function digicode_install() {

}

function digicode_update() {

}


function digicode_remove() {

}

?>


desktop/php/digicode.php
<?php
if (!isConnect('admin')) {
    throw new Exception('{{401 - Accès non autorisé}}');
}
$plugin = plugin::byId('digicode');
sendVarToJS('eqType', $plugin->getId());
$eqLogics = eqLogic::byType($plugin->getId());
?>

<div class="row row-overflow">
    <div class="col-lg-2 col-md-3 col-sm-4">
        <div class="bs-sidebar">
            <ul id="ul_eqLogic" class="nav nav-list bs-sidenav">
                <a class="btn btn-default eqLogicAction" style="width : 100%;margin-top : 5px;margin-bottom: 5px;" data-action="add"><i class="fa fa-plus-circle"></i> {{Ajouter un digicode}}</a>
                <li class="filter" style="margin-bottom: 5px;"><input class="filter form-control input-sm" placeholder="{{Rechercher}}" style="width: 100%"/></li>
                <?php
                foreach ($eqLogics as $eqLogic) {
                    $opacity = ($eqLogic->getIsEnable()) ? '' : jeedom::getConfiguration('eqLogic:style:noactive');
                    echo '<li class="cursor li_eqLogic" data-eqLogic_id="' . $eqLogic->getId() . '" style="' . $opacity .'"><a>' . $eqLogic->getHumanName(true) . '</a></li>';
                }
                ?>
            </ul>
        </div>
    </div>

    <div class="col-lg-10 col-md-9 col-sm-8 eqLogicThumbnailDisplay" style="border-left: solid 1px #EEE; padding-left: 25px;">
        <legend>{{Mes digicode}}</legend>
        <legend><i class="fa fa-cog"></i>  {{Gestion}}</legend>
        <div class="eqLogicThumbnailContainer">
            <div class="cursor eqLogicAction" data-action="add" style="text-align: center; background-color : #ffffff; height : 120px;margin-bottom : 10px;padding : 5px;border-radius: 2px;width : 160px;margin-left : 10px;" >
                <i class="fa fa-plus-circle" style="font-size : 6em;color:#94ca02;"></i>
                <br>
                <span style="font-size : 1.1em;position:relative; top : 23px;word-break: break-all;white-space: pre-wrap;word-wrap: break-word;color:#94ca02">{{Ajouter}}</span>
            </div>
            <div class="cursor eqLogicAction" data-action="gotoPluginConf" style="text-align: center; background-color : #ffffff; height : 120px;margin-bottom : 10px;padding : 5px;border-radius: 2px;width : 160px;margin-left : 10px;">
                <i class="fa fa-wrench" style="font-size : 6em;color:#767676;"></i>
                <br>
                <span style="font-size : 1.1em;position:relative; top : 15px;word-break: break-all;white-space: pre-wrap;word-wrap: break-word;color:#767676">{{Configuration}}</span>
            </div>
        </div>
        <legend><i class="fa fa-table"></i> {{Mes digicodes}}</legend>
        <div class="eqLogicThumbnailContainer">
            <?php
            foreach ($eqLogics as $eqLogic) {
                $opacity = ($eqLogic->getIsEnable()) ? '' : jeedom::getConfiguration('eqLogic:style:noactive');
                echo '<div class="eqLogicDisplayCard cursor" data-eqLogic_id="' . $eqLogic->getId() . '" style="text-align: center; background-color : #ffffff; height : 200px;margin-bottom : 10px;padding : 5px;border-radius: 2px;width : 160px;margin-left : 10px;' . $opacity . '" >';
                echo '<img src="' . $plugin->getPathImgIcon() . '" height="105" width="95" />';
                echo "<br>";
                echo '<span style="font-size : 1.1em;position:relative; top : 15px;word-break: break-all;white-space: pre-wrap;word-wrap: break-word;">' . $eqLogic->getHumanName(true, true) . '</span>';
                echo '</div>';
            }
            ?>
        </div>
    </div>

    <div class="col-lg-10 col-md-9 col-sm-8 eqLogic" style="border-left: solid 1px #EEE; padding-left: 25px;display: none;">
        <a class="btn btn-success eqLogicAction pull-right" data-action="save"><i class="fa fa-check-circle"></i> {{Sauvegarder}}</a>
        <a class="btn btn-danger eqLogicAction pull-right" data-action="remove"><i class="fa fa-minus-circle"></i> {{Supprimer}}</a>
        <a class="btn btn-default eqLogicAction pull-right" data-action="configure"><i class="fa fa-cogs"></i> {{Configuration avancée}}</a>
        <ul class="nav nav-tabs" role="tablist">
            <li role="presentation"><a href="#" class="eqLogicAction" aria-controls="home" role="tab" data-toggle="tab" data-action="returnToThumbnailDisplay"><i class="fa fa-arrow-circle-left"></i></a></li>
            <li role="presentation" class="active"><a href="#eqlogictab" aria-controls="home" role="tab" data-toggle="tab"><i class="fa fa-tachometer"></i> {{Equipement}}</a></li>
            <li role="presentation"><a href="#commandtab" aria-controls="profile" role="tab" data-toggle="tab"><i class="fa fa-list-alt"></i> {{Commandes}}</a></li>
        </ul>
        <div class="tab-content" style="height:calc(100% - 50px);overflow:auto;overflow-x: hidden;">
            <div role="tabpanel" class="tab-pane active" id="eqlogictab">
                <br/>
                <form class="form-horizontal">
                    <fieldset>
                        <div class="form-group">
                            <label class="col-sm-3 control-label">{{Nom de l'équipement template}}</label>
                            <div class="col-sm-3">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="id" style="display : none;" />
                                <input type="text" class="eqLogicAttr form-control" data-l1key="name" placeholder="{{Nom de l'équipement template}}"/>
                            </div>
                        </div>
                        <div class="form-group">
                            <label class="col-sm-3 control-label" >{{Objet parent}}</label>
                            <div class="col-sm-3">
                                <select id="sel_object" class="eqLogicAttr form-control" data-l1key="object_id">
                                    <option value="">{{Aucun}}</option>
                                    <?php
                                    foreach (object::all() as $object) {
                                        echo '<option value="' . $object->getId() . '">' . $object->getName() . '</option>';
                                    }
                                    ?>
                                </select>
                            </div>
                        </div>
                        <div class="form-group">
                            <label class="col-sm-3 control-label">{{Catégorie}}</label>
                            <div class="col-sm-9">
                                <?php
                                foreach (jeedom::getConfiguration('eqLogic:category') as $key => $value) {
                                    echo '<label class="checkbox-inline">';
                                    echo '<input type="checkbox" class="eqLogicAttr" data-l1key="category" data-l2key="' . $key . '" />' . $value['name'];
                                    echo '</label>';
                                }
                                ?>
                            </div>
                        </div>
                        <div class="form-group">
                            <label class="col-sm-3 control-label"></label>
                            <div class="col-sm-9">
                                <label class="checkbox-inline"><input type="checkbox" class="eqLogicAttr" data-l1key="isEnable" checked/>{{Activer}}</label>
                                <label class="checkbox-inline"><input type="checkbox" class="eqLogicAttr" data-l1key="isVisible" checked/>{{Visible}}</label>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Statut de l'alarme}}</label>
                        <div class="col-sm-6">
                            <div class="input-group">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="configuration" data-l2key="digicodeCmdStatus"/>
                                <span class="input-group-btn">
                                    <a class="btn btn-default listCmdInfo"><i class="fa fa-list-alt"></i></a>
                                </span>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Activation mode total}}</label>
                        <div class="col-sm-6">
                            <div class="input-group">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="configuration" data-l2key="digicodeCmdTotal"/>
                                <span class="input-group-btn">
                                    <a class="btn btn-default listCmdActionOther"><i class="fa fa-list-alt"></i></a>
                                </span>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Activation mode Partiel}}</label>
                        <div class="col-sm-6">
                            <div class="input-group">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="configuration" data-l2key="digicodeCmdPartiel"/>
                                <span class="input-group-btn">
                                    <a class="btn btn-default listCmdActionOther"><i class="fa fa-list-alt"></i></a>
                                </span>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Désactivation alarme}}</label>
                        <div class="col-sm-6">
                            <div class="input-group">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="configuration" data-l2key="digicodeCmdDesactivation"/>
                                <span class="input-group-btn">
                                    <a class="btn btn-default listCmdActionOther"><i class="fa fa-list-alt"></i></a>
                                </span>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Etat des portes}}</label>
                        <div class="col-sm-6">
                            <div class="input-group">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="configuration" data-l2key="digicodeEtatPortes"/>
                                <span class="input-group-btn">
                                    <a class="btn btn-default listCmdInfo"><i class="fa fa-list-alt"></i></a>
                                </span>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Etat des fenêtres}}</label>
                        <div class="col-sm-6">
                            <div class="input-group">
                                <input type="text" class="eqLogicAttr form-control" data-l1key="configuration" data-l2key="digicodeEtatFenetres"/>
                                <span class="input-group-btn">
                                    <a class="btn btn-default listCmdInfo"><i class="fa fa-list-alt"></i></a>
                                </span>
                            </div>
                        </div>
                        <label class="col-sm-4 control-label">{{Délais}}</label>
                        <div class="col-sm-6">
                            <input type="text" class="eqLogicAttr configuration form-control" data-l1key="configuration" data-l2key="digicodeDelais" />
                        </div>


                </fieldset>
            </form>
        </div>



        <div role="tabpanel" class="tab-pane" id="commandtab">
            <a class="btn btn-success btn-sm cmdAction pull-right" data-action="add" style="margin-top:5px;"><i class="fa fa-plus-circle"></i> {{Commandes}}</a><br/><br/>
            <table id="table_cmd" class="table table-bordered table-condensed">
                <thead>
                    <tr>
                        <th>{{Nom}}</th><th>{{Code}}</th><th>{{Type}}</th><th>{{Action}}</th>
                    </tr>
                </thead>
                <tbody>
                </tbody>
            </table>
        </div>
    </div>
</div>
</div>

<?php include_file('desktop', 'digicode', 'js', 'digicode');?>
<?php include_file('core', 'plugin.template', 'js');?>


core/ajax/digicode.ajax.php
<?php

/* This file is part of Jeedom.
*
* Jeedom is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Jeedom is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Jeedom. If not, see <http://www.gnu.org/licenses/>.
*/

try {
  require_once dirname(__FILE__) . '/../../../../core/php/core.inc.php';
  include_file('core', 'authentification', 'php');

  if (!isConnect('admin')) {
    throw new Exception(__('401 - Accès non autorisé', __FILE__));
  }

  switch (init('action')) {
    case 'Modify':
      $id  = init('id');
      $cmdid = init('cmdid');
      $code 	= init('code');
      $cmdName = init('cmdName');
      ajax::success(digicode::modifyUserCode($id,$cmdid,$code,$cmdName));
      break;

      case 'Remove':
      $id  = init('id');
      $cmdid = init('cmdid');
      $cmdName = init('cmdName');
      ajax::success(digicode::RemoveUser($id,$cmdid,$cmdName));
      break;
      case 'Add':
      $id  = init('id');
      $user = init('user');
      $code 	= init('code');
      ajax::success(digicode::AddUser($id,$user,$code));
      break;
    }

    throw new Exception(__('Aucune méthode correspondante à : ', __FILE__) . init('action'));
    /*     * *********Catch exeption*************** */
  } catch (Exception $e) {
    ajax::error(displayException($e), $e->getCode());
  }


core/class/digicode.class.php
<?php

require_once dirname(__FILE__).'/../../../../core/php/core.inc.php';

class digicode extends eqLogic {

  public static function modifyUserCode($id,$cmdid,$code,$cmdName)
  {
    $digicode    = eqLogic::byId($id);
    $cmd = new digicodeCmd();
    $cmd->setName($cmdName);
    $cmd->setid($cmdid);
    $cmd->setEqLogic_id($id);
    $cmd->setType('info');
    $cmd->setSubType('numeric');
    $cmd->setConfiguration('userCode', $code);
    $cmd->save();
  }

  public static function RemoveUser($id,$cmdid,$cmdName)
  {
    $cmd = cmd::byId($cmdid);
    if (!is_object($cmd)) {
      throw new Exception('User ID inconnu');
    }
    $cmd->remove();

  }

  public static function AddUser($id,$user,$code)
  {
    $digicode    = eqLogic::byId($id);
    $digicodeCmd = $digicode->getCmd(null, $user);
    if (!is_object($digicodeCmd)) {
      $digicodeCmd = new digicodeCmd();
      $digicodeCmd->setName(__($user, __FILE__));
      $digicodeCmd->setEqLogic_id($digicode->getId());
      $digicodeCmd->setType('info');
      $digicodeCmd->setSubType('numeric');
    }
	$digicodeCmd->setOrder(99);
    $digicodeCmd->setConfiguration('userCode', $code);
    $digicodeCmd->save();
  }

  public static function checkandactivate($mode,$eqLogic)
  {
    $etatAlarme =  cmd::byString($eqLogic->getConfiguration('digicodeCmdStatus'));
    $valueEtatAlarme  = $etatAlarme->execCmd();
    $etatFenetres = cmd::byString($eqLogic->getConfiguration('digicodeEtatFenetres'));
    $valueEtatFenetres  = $etatFenetres->execCmd();
    $etatPortes = cmd::byString($eqLogic->getConfiguration('digicodeEtatPortes'));
    $valueEtatPorte  = $etatPortes->execCmd();


    if($valueEtatFenetres == 1 && $valueEtatPorte == 0){
      log::add('digicode', 'DEBUG', 'Au moins une fenêtre est restée ouverte !');
      $eqLogic->checkAndUpdateCmd('message', 'Fenetre(s) ouverte(s)');
      $eqLogic->save();
    }elseif($valueEtatPorte == 1 && $valueEtatFenetres == 0){
      log::add('digicode', 'DEBUG', 'Au moins une porte est restée ouverte !');
      $eqLogic->checkAndUpdateCmd('message', 'Porte(s) ouverte(s)');
      $eqLogic->save();
    }elseif($valueEtatPorte == 1 && $valueEtatFenetres == 1){
      log::add('digicode', 'DEBUG', 'Au moins une porte est restée ouverte !');
      $eqLogic->checkAndUpdateCmd('message', 'Porte(s) et Fenêtre(s) ouverte(s)');
      $eqLogic->save();
    }elseif($valueEtatAlarme ==1){
      log::add('digicode', 'DEBUG', 'Alarme déjà active !');
      $eqLogic->checkAndUpdateCmd('message', 'Alarme déjà active');
      $eqLogic->save();
    }else{
      log::add('digicode', 'DEBUG', 'Tout est ok pour l\'activation');
      $eqLogic->checkAndUpdateCmd('message', '');
      $eqLogic->save();
      return true;
    }

  }


  public function postUpdate()
  {
    $digicodeCmd = $this->getCmd(null, 'exec alarme');
    if (!is_object($digicodeCmd)) {
      $digicodeCmd = new digicodeCmd();
      $digicodeCmd->setName(__('exec alarme', __FILE__));
      $digicodeCmd->setEqLogic_id($this->getId());
      $digicodeCmd->setIsVisible(0);
      $digicodeCmd->setLogicalId('exec alarme');
      $digicodeCmd->setType('action');
      $digicodeCmd->setSubType('message');

    }
    $digicodeCmd->setConfiguration('type', 'cmdwiget');
    $digicodeCmd->save();

    $digicodeCmd = $this->getCmd(null, 'save code');
    if (!is_object($digicodeCmd)) {
      $digicodeCmd = new digicodeCmd();
      $digicodeCmd->setName(__('save code', __FILE__));
      $digicodeCmd->setEqLogic_id($this->getId());
      $digicodeCmd->setIsVisible(0);
      $digicodeCmd->setLogicalId('save code');
      $digicodeCmd->setType('action');
      $digicodeCmd->setSubType('message');
    }
    $digicodeCmd->setConfiguration('type', 'cmdwiget');
    $digicodeCmd->save();

    $digicodeCmd = $this->getCmd(null, 'etat');
    if (!is_object($digicodeCmd)) {
      $digicodeCmd = new digicodeCmd();
      $digicodeCmd->setName(__('etat', __FILE__));
      $digicodeCmd->setEqLogic_id($this->getId());
      $digicodeCmd->setIsVisible(0);
      $digicodeCmd->setLogicalId('etat');
      $digicodeCmd->setType('info');
      $digicodeCmd->setSubType('numeric');
    }
    $digicodeCmd->setConfiguration('type', 'cmdwiget');
    $digicodeCmd->save();

    $digicodeCmd = $this->getCmd(null, 'message');
    if (!is_object($digicodeCmd)) {
      $digicodeCmd = new digicodeCmd();
      $digicodeCmd->setName(__('message', __FILE__));
      $digicodeCmd->setEqLogic_id($this->getId());
      $digicodeCmd->setIsVisible(0);
      $digicodeCmd->setLogicalId('message');
      $digicodeCmd->setType('info');
      $digicodeCmd->setSubType('string');
    }
    $digicodeCmd->setConfiguration('type', 'cmdwiget');
    $digicodeCmd->save();

    /********** Getters and setters **********/

  }


  public function toHtml($_version = 'dashboard')
  {
    $replace = $this->preToHtml($_version);
    if (!is_array($replace)) {
      return $replace;
    }
    $version = jeedom::versionAlias($_version);
    foreach ($this->getCmd('info') as $cmd) {
      $replace['#' . $cmd->getLogicalId() . '#']    = $cmd->execCmd();
      $replace['#' . $cmd->getLogicalId() . '_id#'] = $cmd->getId();
      if ($cmd->getIsHistorized() == 1) {
        $replace['#' . $cmd->getLogicalId() . '_history#'] = 'history cursor';
      }
    }

    foreach ($this->getCmd('action') as $cmd) {
      $replace['#' . $cmd->getLogicalId() . '_id#'] = $cmd->getId();
    }

    $cmdEtatFenetres = cmd::byString($this->getConfiguration('digicodeEtatFenetres'));
    $valueEtatFenetres  = $cmdEtatFenetres->execCmd();
    $replace['#EtatFenetres#'] = $valueEtatFenetres;
    $cmdEtatPortes = cmd::byString($this->getConfiguration('digicodeEtatPortes'));
    $valueEtatPortes  = $cmdEtatPortes->execCmd();
    $replace['#EtatPortes#'] = $valueEtatPortes;
    $valuedigicodeDelais = $this->getConfiguration('digicodeDelais');
    $replace['#digicodeDelais#'] = $valuedigicodeDelais;


    return $this->postToHtml($_version, template_replace($replace, getTemplate('core', $version, 'digicode', 'digicode')));
  }



}
class digicodeCmd extends cmd {


  public function execute($_options = null)
  {
    $eqLogic = $this->getEqLogic();
    $eqLogics = eqLogic::byType('digicode');
    $cmdAlarmeDesactive = $eqLogic->getConfiguration('digicodeCmdDesactivation');

    switch ($this->getLogicalId()) {

      case 'save code':
      $value    = $_options['message'];
      $mode   = substr($value, -1, 1);
      $code    = substr($value, 0, 4);

      foreach ($eqLogics as $digicode) {
        if ($digicode->getIsEnable() == 1) {
          foreach ($digicode->getCmd('info') as $cmd) {

            if (!empty($cmd->getConfiguration('userCode')) && $cmd->getConfiguration('userCode') == $code){
              switch ($mode) {
                case 'P':
                $returnCheck = $eqLogic->checkandactivate($mode,$eqLogic);
                if( $returnCheck ==1){
                  log::add('digicode', 'DEBUG', 'Activation de l\'alarme partielle par l\'utilisateur '. $cmd->getName() . ' (Code utilisé ' .$cmd->getConfiguration('userCode') .')' );
                  return 2;
                }
                break;

                case 'T':
                $returnCheck =$eqLogic->checkandactivate($mode,$eqLogic);
                if( $returnCheck ==1){
                  log::add('digicode', 'DEBUG', 'Activation de l\'alarme totale par l\'utilisateur '. $cmd->getName() . ' (Code utilisé ' .$cmd->getConfiguration('userCode') .')' );
                  return 1;
                }
                break;

                case 'D':
                log::add('digicode', 'DEBUG', 'Désactivation de l\'alarme par l\'utilisateur '. $cmd->getName() . ' (Code utilisé ' .$cmd->getConfiguration('userCode') .')' );
                $cmd2 = cmd::byString($cmdAlarmeDesactive);
                $cmd2->execCmd();
                $eqLogic->checkAndUpdateCmd('etat', '0');
                $eqLogic->save();
                $eqLogic->checkAndUpdateCmd('message', '');
                $eqLogic->save();
                return 0;
                break;

              }
            } elseif (!empty($cmd->getConfiguration('userCode')) && $cmd->getConfiguration('userCode') != $code){
              log::add('digicode', 'DEBUG', 'Code erroné');
              $eqLogic->checkAndUpdateCmd('message', 'Code erroné');
              $eqLogic->save();
            }
        }
        }
      }
      break;

      case 'exec alarme':
      $value    = $_options['message'];
      $eqLogic = $this->getEqLogic();
      $cmdAlarmetotal = $eqLogic->getConfiguration('digicodeCmdTotal');
      $cmdAlarmePartiel = $eqLogic->getConfiguration('digicodeCmdPartiel');

      if($value == 1){
        $cmd = cmd::byString($cmdAlarmetotal);
        $cmd->execCmd();
        $eqLogic->checkAndUpdateCmd('etat', '1');
        $eqLogic->save();
      }
      if($value == 2){

        $cmd = cmd::byString($cmdAlarmePartiel);
        $cmd->execCmd();
        $eqLogic->checkAndUpdateCmd('etat', '2');
        $eqLogic->save();
      }
      break;
    }

}
/********** Getters and setters **********/

}


core/template/dashboard/css/digicode.js
jQuery.fn.extend({
  
  digicode: function(callback, delayReset) {
    
    this.html('<ul class="digiEvent"><li></li><li></li><li></li><li></li><li></li></ul><ul class="digicodePanel"><div> <li>1</li><li>2</li><li>3</li><li class="digiFunction digiFunctionPartiel">P</li></div><div> <li>4</li><li>5</li><li>6</li><li class="digiFunction digiFunctionTotal">T</li></div><div> <li>7</li><li>8</li><li>9</li><li class="digiFunction digiFunctionDesarmer">D</li></div><div> <li>A</li><li>0</li><li>B</li><li class="digiReset">X</li></div></ul>');
    this.addClass('digicode');   
    this.delayReset = $.isNumeric(delayReset) ? delayReset : 4; //DELAY EN SECONDE POUR AUTO RESET CODE
    this.jeedomExecute = $.isFunction(callback) ? callback : (function () {}) ; 	
    this.keys = this.find('.digicodePanel li');
    this.Displays = this.find('.digiEvent li');
    this.inputs = [];
    this.timer = null;
    this.delayReset *= 1000;

    this.displayInputs = (function() {
      this.Displays.removeClass('digiFilled digiFilledOK');
      $.each(this.inputs, (function(i, e) {
        this.Displays.eq(i).addClass('digiFilled');
      }).bind(this));
    }).bind(this);
    
    this.clearCode = (function() {
      this.inputs = [];
      this.displayInputs();
      clearInterval(this.timer);
    }).bind(this);

    this.resetTimer = (function(resetTimer) {
      if (this.timer != null) {
        clearInterval(this.timer);
      }
      this.timer = setInterval(this.clearCode, this.delayReset);
    }).bind(this);
   
    this.codeReady = (function() {
      this.jeedomExecute(this.inputs.join(''));
        setTimeout((function() {
          this.Displays.addClass('digiFilledOK');
        }).bind(this), 200);
        setTimeout((function() {
          this.clearCode();
        }).bind(this), 500);
    }).bind(this);

    this.keys.on(('ontouchstart' in document.documentElement) ? "touchstart" : "click", (function(e) {
      var el = $(e.currentTarget);
      if (el.hasClass('digiReset')) {
        this.clearCode();
      }
      else {
        el.addClass('digiSel');
        this.inputs.push(el.text());
        this.displayInputs();
        this.resetTimer();
        if (this.inputs.length == 5) {
          this.codeReady();
        }
      }
    }).bind(this));

    this.keys.on('mouseup mouseleave touchend', function() {
      var el = $(this);          
      if (!el.hasClass('digiReset')) {
        setTimeout(function() { 
          el.removeClass('digiSel');
        }, 150);
      }
    });
        
  }
});

desktop/modal/modal.digicode.php
<?php

/* This file is part of Jeedom.
*
* Jeedom is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Jeedom is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Jeedom. If not, see <http://www.gnu.org/licenses/>.
*/

if (!isConnect()) {
    throw new Exception('{{401 - Accès non autorisé}}');
}
if (init('id') == '') {
    throw new Exception(__('L\'id ne peut etre vide', __FILE__));
}
$digicode = digicode::byId(init('id'));
if (!is_object($digicode)) {
    throw new Exception(__('L\'équipement est introuvable : ', __FILE__) . init('id'));
}
if ($digicode->getEqType_name() != 'digicode') {
    throw new Exception(__('Cet équipement n\'est pas de type motion : ', __FILE__) . $digicode->getEqType_name());
}
$id = init('id');
sendVarToJS('id', $id);
?>
<?php
echo '<table class="table table-bordered table-condensed" style="width: 100%">';
echo '    <thead>';
echo '        </tr>';
echo '            <th style="width: 40%">{{Utilisateur}}</th><th style="width: 35%">{{Code}}</th><th style="width: 25%"></th>';
echo '        </tr>';

foreach ($digicode->getCmd('info') as $cmd) {
    if($cmd->getName() != 'etat' && $cmd->getName() != 'message'){

        echo      '<tr class="cmd" data-cmd_id="' . $cmd->getname() . '">';
        echo '<td>';
        echo $cmd->getname();
        echo '</td>';
        echo '<td>';
        echo $cmd->getConfiguration('userCode');
        echo '</td>';
        echo '<td>';
        echo ' <a class="btn btn-success bt_Modify btn-xs" data-id="' . $cmd->getid() . '"><i class="fa fa-gear"></i></a></center>';
        echo ' <a class="btn btn-danger bt_Remove btn-xs" data-id="' . $cmd->getid() . '"><i class="fa fa-trash-o"></i></a></center>';
        echo '</td>';
        echo '</tr>';
    }
}

echo '    </thead>';
echo '</table>';
echo ' <div align="right">';
echo ' <a class="btn btn-success bt_Add btn-xs">Ajouter</a> ';
echo ' <a class="btn btn-danger bt_Cancel btn-xs">Annuler</a> ';
echo ' </div>';

?>

<script>

$('.bt_Cancel').on('click', function() {
    $('#md_modal').dialog("close");
});
$('.bt_Modify').on('click', function() {
    var cmdid = $(this).attr('data-id');
    var cmdName = $(this).closest('.cmd').attr('data-cmd_id');

    bootbox.prompt("Nouveau code", function(code){

        var code = code;
        action = 'Modify';
        $.ajax({// fonction permettant de faire de l'ajax
        type: "POST", // methode de transmission des données au fichier php
        url: "plugins/digicode/core/ajax/digicode.ajax.php", // url du fichier php
        data: {
            action: action,
            id: id,
            cmdid:cmdid,
            cmdName:cmdName,
            code: code,
        },
        dataType: 'json',
        error: function (request, status, error) {
            handleAjaxError(request, status, error);
        },
        success: function(data) {
            $('#md_modal').dialog("close");

            if (data.state != 'ok') {
                $('#div_alert').showAlert({message: data.result, level: 'danger'});
                return;
            }
        }
    });
});
});

$('.bt_Remove').on('click', function() {
    var cmdid = $(this).attr('data-id');
    var cmdName = $(this).closest('.cmd').attr('data-cmd_id');
    action = 'Remove';
    $.ajax({// fonction permettant de faire de l'ajax
    type: "POST", // methode de transmission des données au fichier php
    url: "plugins/digicode/core/ajax/digicode.ajax.php", // url du fichier php
    data: {
        action: action,
        id: id,
        cmdid:cmdid,
        cmdName:cmdName,
    },
    dataType: 'json',
    error: function (request, status, error) {
        handleAjaxError(request, status, error);
    },
    success: function(data) { // si l'appel a bien fonctionné
    if (data.state != 'ok') {
        $('#div_alert').showAlert({message: data.result, level: 'danger'});
        return;
    }
    $('#md_modal').dialog("close");

}
});
});

$('.bt_Add').on('click', function() {
    var cmdid = $(this).attr('data-id');
    var cmdName = $(this).closest('.cmd').attr('data-cmd_id');
    // var url = "index.php?";
    bootbox.prompt("Utilisateur", function(user){

        var code = code;
        bootbox.prompt("Code", function(code){
            action = 'Add';
            $.ajax({// fonction permettant de faire de l'ajax
            type: "POST", // methode de transmission des données au fichier php
            url: "plugins/digicode/core/ajax/digicode.ajax.php", // url du fichier php
            data: {
                action: action,
                id: id,
                user:user,
                code: code,
            },
            dataType: 'json',
            error: function (request, status, error) {
                handleAjaxError(request, status, error);
            },
            success: function(data) {
                $('#md_modal').dialog("close")

                if (data.state != 'ok') {
                    $('#div_alert').showAlert({message: data.result, level: 'danger'});
                    return;
                }
            }
        });
    });
});
});

</script>

<?php include_file('desktop', 'digicode', 'js', 'digicode');?>
<?php include_file('core', 'plugin.template', 'js');?>
